Multiplexing token filter

romseygeek opened the pull request:
This adds a `multiplexer` token filter to elasticsearch, which allows you to run tokens through multiple different tokenfilters and stack the results.  For example, you can now easily index the original form of a token, its lowercase form and a stemmed form all at the same position, allowing you to search for stemmed and unstemmed tokens in the same field.

elasticmachine commented:
Pinging @elastic/es-search-aggs

romseygeek commented:
This would solve the usecase mentioned in https://github.com/elastic/elasticsearch/issues/22478

jpountz commented on docs/reference/analysis/tokenfilters/multiplexer-tokenfilter.asciidoc:
```diff
@@ -0,0 +1,110 @@
+            "filter" : {
+                "my_multiplexer" : {
+                    "type" : "multiplexer",
+                    "filters" : [ "identity", "lowercase", "lowercase, porter_stem" ]
```
thinking about the API I wonder whether this should be an array of arrays, eg. `[ [], [ "lowercase" ], [ "lowercase", "porter_stem" ] ]`.

jpountz replied:
then maybe make squares optional for single-element lists?
`[ [], "lowercase", [ "lowercase", "porter_stem" ] ]`

romseygeek replied:
How easy is that to do with the Settings API?  I've had a quick look, and it seems to expect either single keys or lists of strings, whereas here we'd need a list of variant types.

jpountz replied:
Ok to keep a comma-separated list then. The other thing that worried me is what happens if the user also has a filter whose name is `identity`. Even though I like it a bit less from an API perspective, maybe it would be better to add a `preserve_original` setting or something like that to the `multiplexer` type so that we do not have to add `identity` as a reserved filter name.

romseygeek commented:
> Out of curiosity does it work well with filters that modify the graph like stop or synonym filters?

The multiplexing filter will only pass one token at a time to its child filters, so filters that need to read ahead, like shingle or synonym filters, won't work.  A stop filter would work, in that if you chained a stop filter and then a stemmer, the stop filter could prevent terms being sent to the stemmer.

martijnvg commented on server/src/main/java/org/elasticsearch/index/analysis/MultiplexingTokenFilterFactory.java:
```diff
@@ -0,0 +1,177 @@
+package org.elasticsearch.index.analysis;
```
I think we should move this token filter and its tests to the `org.elasticsearch.analysis.common` package in the `analysis-common` module with all the other migrated analysis components.

Instead of wiring this token filter up in `AnalysisRegistry` you will need to do that in `CommonAnalysisPlugin`.

martijnvg commented on server/src/test/java/org/elasticsearch/index/analysis/AnalysisRegistryTests.java:
```diff
@@ -239,4 +243,126 @@ public void testEnsureCloseInvocationProperlyDelegated() throws IOException {
         verify(mock).close();
     }
+
+    private final class TruncateTokenFilter extends TokenFilter {
```
Maybe use Lucene's `TruncateTokenFilter` / `UpperCaseFilter`?

romseygeek commented:
I moved things to the common analysis plugin.  I still need to do a bit of wiring in `AnalysisRegistry` to pass the list of registered filters back to the multiplexer so it can resolve things properly.  I added a new interface for this, but another alternative would be to add a method to `TokenFilterFactory` with a default no-op implementation.

martijnvg commented on server/src/main/java/org/elasticsearch/index/analysis/ReferringFilterFactory.java:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * Called with a map of all registered filter factories
+     */
+    void addReferences(Map<String, TokenFilterFactory> factories);
```
This should be invoked once per token filter factory? In that case I would rename this method to `setReference(...)` to highlight this.

jpountz commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/MultiplexingTokenFilterFactory.java:
```diff
@@ -0,0 +1,200 @@
+            @Override
+            public void reset() throws IOException {
+                super.reset();
+                selector = 0;
```
should it set `selector = filterCount - 1` and merge the two if statements from `incrementToken`?

jpountz commented:
> The multiplexing filter will only pass one token at a time to its child filters, so filters that need to read ahead, like shingle or synonym filters, won't work.

Let's document this?

nik9000 commented on docs/reference/analysis/tokenfilters/multiplexer-tokenfilter.asciidoc:
```diff
@@ -0,0 +1,114 @@
+A token filter of type `multiplexer` will emit multiple tokens at the same position,
+each version of the token having been run through a different filter.
+
+Note that the child filters will in effect be passed a mock tokenstream consisting
```
I might put this as a `WARNING: ` in the `Options` section. And I'd flip the sentence around to something like "Shingle or multi-word synonym token filters will not function normally when they are declared in the `filters` array because they read ahead internally which is unsupported by the `multiplexer`.

nik9000 commented on docs/reference/analysis/tokenfilters/multiplexer-tokenfilter.asciidoc:
```diff
@@ -0,0 +1,114 @@
+
+[source,js]
+--------------------------------------------------
+POST /multiplexer_example/_analyze
```
This is seriously one of my favorite APIs in Elasticsearch.

nik9000 commented on docs/reference/analysis/tokenfilters/multiplexer-tokenfilter.asciidoc:
```diff
@@ -0,0 +1,114 @@
+      "position": 2
+    },
+    {
+      "token": "home",
```
Hmmmm. It might be nice to add a callout (looks like `<1>`) to explain the duplicate. I presume we get the duplicate because we don't deduplicate.

romseygeek replied:
I was going to add a deduplication step as well to remove this confusion, but then I noticed that we don't seem to have the deduplication filter exposed in ES anywhere.  I'll open another issue for that, as I think it will be very useful combined with a multiplexer

jpountz replied:
I wonder that we might want to remove duplicates by default (or even enforce it). Otherwise eg. terms that are not modified through lowercasing or stemming will artificially get higher term freqs?

nik9000 replied:
I'd prefer to to document the duplicates and have folks use the deduplicating filter that @romseygeek proposed today. I like documenting it so folks understand what costs they are paying.

Also, if the token steam comes with duplicates on the way into this token filter then adding deduplicating filter by default would deduplicate the existing duplicates as a side effect, right?

jpountz replied:
Sorry, I just saw this message. I don't think the documentation path gives the best user experience as I can't think of a use-case to retain duplicates if multiple filters produce the same token. That said, I agree that simply adding a deduplicating token filter feels wrong if the original stream has duplicates, so maybe this is something that needs to be implemented directly into this new token filter.

nik9000 commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/MultiplexingTokenFilterFactory.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    @Override
+    public void setReferences(Map<String, TokenFilterFactory> factories) {
```
I wonder if It'd be cleaner if `create` took the `Map`? It'd certainly make the change larger though.

nik9000 replied:
Maybe a thing to do in a followup because its large but almost entirely mechanical.

romseygeek replied:
It would also be backwards breaking, so would have to be a master-only change, I think?  TokenFilterFactory is an API you're encouraged to use via analysis plugins

nik9000 replied:
We have a special tag for things that break the plugin or java client APIs: "breaking-java". We allow ourselves to use it much more liberally. One day we will have a properly semver-ed plugin API but we don't have one now.

jpountz commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/MultiplexingTokenFilterFactory.java:
```diff
@@ -0,0 +1,200 @@
+    public MultiplexingTokenFilterFactory(IndexSettings indexSettings, Environment env, String name, Settings settings) throws IOException {
+        super(indexSettings, name, settings);
+        this.filterNames = settings.getAsList("filters");
+        this.preserveOriginal = settings.getAsBoolean("preserveOriginal", true);
```
use underscore case instead?
